### ISSUE TITLE
Question: How to get the name (as string) from a pointer-to-base?

### ISSUE THREAD

**Issue by globberwops** [BlackMATov/meta.hpp#102]
Hi Matvey,

I just cannot seem to figure out an easy way of getting the name from a pointer-to-base.

Suppose I have a class hierarchy, all with `META_HPP_ENABLE_POLY_INFO`.
All classes are registered in the `static_scope_` via `typedef_`.

Now somewhere in my codebase I want to know the name of a class in a scope.
The only knowledge I have is the name of the scope and a pointer-to-base of the class.

Basically what I want to do is something like this:
```cpp
const auto* entity...

const auto foo_scope = meta::resolve_scope("Foo");
const auto entity_type = foo_scope.resolve_type(entity->meta_poly_ptr().get_type());
const auto entity_name = entity_type.get_name();
```

What I'm looking for is a replacement of `ponder::UserObject::getClass::name()`.

Cheers,
Martin

**Comment by BlackMATov** [comment]
There are two ways to do it:

1. add names to metadata of classes

```cpp
namespace meta = meta_hpp;
using namespace std::literals;

meta::class_<shape>(meta::metadata_()
    ("name", "shape"s));

meta::class_<rectangle>(meta::metadata_()
    ("name", "rectangle"s));

std::unique_ptr<shape> s = std::make_unique<rectangle>();

meta::pointer_type most_derived_pointer_type = s->meta_poly_ptr().get_type().as_pointer();
CHECK(most_derived_pointer_type == meta::resolve_type<rectangle*>());

meta::class_type most_derived_class_type = most_derived_pointer_type.get_data_type().as_class();
CHECK(most_derived_class_type == meta::resolve_type<rectangle>());

CHECK(most_derived_class_type.get_metadata().at("name").as<std::string>() == "rectangle");
```

2. write a function that finds a typedef name in a scope

```cpp
std::string find_typedef_name_in_scope(const meta::scope& scope, const meta::any_type& type) {
    for (const auto& [name, scope_type] : scope.get_typedefs()) {
        if ( type == scope_type ) { return name; }
    }
    return "";
}

std::unique_ptr<shape> s = std::make_unique<rectangle>();

meta::pointer_type most_derived_pointer_type = s->meta_poly_ptr().get_type().as_pointer();
CHECK(most_derived_pointer_type == meta::resolve_type<rectangle*>());

meta::class_type most_derived_class_type = most_derived_pointer_type.get_data_type().as_class();
CHECK(most_derived_class_type == meta::resolve_type<rectangle>());

CHECK(find_typedef_name_in_scope(shapes, most_derived_class_type) == "rectangle");
```

By the way, you can use `meta::resolve_type` to get most derived polymorphic type instead of using `meta_poly_ptr` directly.

```cpp
std::unique_ptr<shape> s = std::make_unique<rectangle>();
meta::class_type most_derived_class_type = meta::resolve_type(*s);
CHECK(most_derived_class_type == meta::resolve_type<rectangle>());
```

P.S.
I think I would prefer the metadata way, because most likely you want the class name, not its alias in the scope, but it's up to you of course!

**Comment by globberwops** [comment]
Thanks again!

I've decided to implement functions for both cases, bc. sometimes I need the exact name of the type, and at other times I need the alias.
